sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala

```diff
@@ -36,7 +36,13 @@ abstract class AverageLike(child: Expression) extends DeclarativeAggregate {
   }
 
   private lazy val sumDataType = child.dataType match {
-    case _ @ DecimalType.Fixed(p, s) => DecimalType.bounded(p + 10, s)
+    /*
+     * In case of sum of decimal ( assuming another decimal of same precision and scale)
+     * Refer : org.apache.spark.sql.catalyst.analysis.DecimalPrecision
+     * Precision : max(s1, s2) + max(p1 - s1, p2 - s2) + 1
+     * Scale : max(s1, s2)
+     */
+    case _ @ DecimalType.Fixed(p, s) => DecimalType.adjustPrecisionScale(s + (p - s) + 1, s)
```

Comment:
The point is that here the sum operation is executed many times, not only once. So I am not sure that this the right way to deal with it. It would be great to check what other RDBMs do in this case.

Reply:
Just checked for sqlserver https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-2017

Reply:
Yes i agree. But the point is arbitrarily having precision increased by 10 can cause loss of scale more often and calculating the times is costly. As avg = (sum(data)/times), can we have precision and scale of sum(data) restricted as described by + operation.?

Reply:
No, we can't because we would risk (well, we would likely hit) an overflow. Indeed, I am not sure if you run all the UTs with your change, but I'd expect many failures due to overflow after this.

Reply:
But division operation will readjust the precision again in average. Can you please give me a example query which can cause overflow as you explained.?

Reply:
well, in your example, with input data of decimal(2,1), this "buffer" with your change would be a decimal (3, 1).. If your input data contains 21 9.1 items, this would overflow (191.1 doesn't fit a decimal(3,1)).

Reply:
well i tested as per your suggestion with my PR : 
```
 sql("create table if not exists table1(salary decimal(2,1))")
 (1 to 22).foreach(_ => sql("insert into table1 values(9.1)"))
 sql("select avg(salary) from table1").show(false)

+-----------+
|avg(salary)|
+-----------+
|9.10000    |
+-----------+
```
which is expected result and i don't see a overflow as divide will readjust precision. Can you test with my patch for a overflow specifically in case of average.?

Reply:
You're right, because we are not checking the overflow in the Add operation, so despite we are in an error condition we don't detect it, but it doesn't sound great to me to rely on a currently missing check. I am not sure, though, if in special cases this can anyway cause an issue also with the current missing check.

Moreover, as you can see from the link I have posted, SQLServer - which is the reference for the way we handle decimals here - uses: decimal(38, s) divided by decimal(10, 0). So SQLServer. So I think this is what we should do eventually, but it implies changing the result type.